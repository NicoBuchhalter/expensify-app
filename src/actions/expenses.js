import uuid from 'uuid';
import firestore from '../firebase/firebase';

const firestoreExpenses = firestore.collection('expenses')

export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
}); 

export const startAddExpense = (expensesData = {}) => {
  return (dispatch) => {
    const {
      description = '', 
      note = '', 
      amount = 0, 
      createdAt = 0       
    } =  expensesData;
    const expense = { description, note, amount, createdAt };
    return firestoreExpenses.add(expense).then((docRef) => {
      dispatch(addExpense({
        id: docRef.id, 
        ...expense
      }));
    });

  };
};

export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch) => {
    return firestoreExpenses.doc(id).delete().then(() => {
      dispatch(removeExpense({ id }));
    });    
  }
};

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});


export const startSetExpenses = () => {
  return (dispatch) => {
    return firestoreExpenses.get().then(querySnapshot => {
      const expenses = querySnapshot.docs.map(docRef => ({
        id: docRef.id,
        ...docRef.data()
      }));
      dispatch(setExpenses(expenses));
    });    
  }
};