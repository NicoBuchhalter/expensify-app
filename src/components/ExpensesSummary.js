import React from 'react';
import { connect } from 'react-redux';
import selectExpensesTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';
import numeral from 'numeral';

export const ExpensesSummary = ({ expenseCount, expensesTotal}) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
  const formattedTotal = numeral(expensesTotal / 100).format('$0,0.00');
  return (
    <div>
      <h1>Viewing {expenseCount} {expenseWord} totalling {formattedTotal}</h1>
    </div>
  );
};

const mapStateToProps = (state) => {
  const expenses = selectExpenses(state.expenses, state.filters)
  return {
    expenseCount: expenses.length,
    expensesTotal: selectExpensesTotal(expenses)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
