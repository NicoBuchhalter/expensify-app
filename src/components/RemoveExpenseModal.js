import React from 'react';
import Modal from 'react-modal';


const RemoveExpenseModal = (props) => (
  <Modal 
    isOpen={props.isModalOpen} 
    contentLabel="Remove Expense"
    closeTimeoutMS={200}
    className="modal"
  >
    <h3 className='modal__title'>{props.expenseDescription}</h3>
    <p className='modal__body'>Are you sure you want to remove this expense?</p>
    <div className='modal__buttons'>
      <button className='button button--secondary' onClick={props.handleRemoveExpense}>Remove</button>
      <button className='button' onClick={props.handleCloseModal}>Cancel</button>
    </div>
  </Modal>
)

export default RemoveExpenseModal;