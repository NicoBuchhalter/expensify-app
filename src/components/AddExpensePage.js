import React from 'react';
import ExpenseForm from './ExpenseForm';
import LoadingPage from './LoadingPage';
import {connect} from 'react-redux';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {

  state = {
    loading: false
  }

  onSubmit = (expense) => {
    this.setState({ loading: true });
    this.props.startAddExpense(expense).then(() => {
      this.props.history.push('/');
    })
  };

  render() {
    return(
      <div>
        { this.state.loading ? 
          (
            <LoadingPage/>
          ) :
          (
            <div>
              <div className='page-header'>
                <div className='content-container'>
                  <h1 className='page-header__title'>Add Expense</h1>
                </div>
              </div>
              <div className='content-container'>
                <ExpenseForm 
                  onSubmit={this.onSubmit}
                  submitText='Save Expense'
                />
              </div>
            </div>
          )
        }
      </div>
    )
  }

}


const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});


export default connect(undefined, mapDispatchToProps)(AddExpensePage);