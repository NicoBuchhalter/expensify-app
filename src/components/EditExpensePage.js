import React from 'react';
import {connect} from 'react-redux';
import RemoveExpenseModal from './RemoveExpenseModal';
import ExpenseForm from './ExpenseForm';
import LoadingPage from './LoadingPage';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {

  state = {
    isModalOpen: false,
    loading: false
  }

  onSubmit = (expense) => {
    this.setState({ loading: true });
    this.props.startEditExpense(this.props.expense.id, expense).then(() => {
      this.props.history.push('/');
    })
  };

  onRemove = () => {
    this.setState({ isModalOpen: false, loading: true });
    this.props.startRemoveExpense(this.props.expense.id).then(() => {
      this.props.history.push('/');
    })
  };

  handleOpenModal = () => {
    this.setState({ isModalOpen: true });
  }

  closeModal = () => {
    this.setState({ isModalOpen: false })
  }

  render() {
    return (
      <div>
        {!this.state.loading ? (
          <div>
            <div className='page-header'>
              <div className='content-container'>
                <h1 className='page-header__title'>Edit Expense</h1>
              </div>
            </div>
            <div className='content-container'>
              <ExpenseForm
                expense={this.props.expense}
                onSubmit={this.onSubmit}
                submitText='Save Expense'
              />
              <button className='button--secondary' onClick={this.handleOpenModal}>Remove Expense</button>
            </div>
            <RemoveExpenseModal 
              isModalOpen={this.state.isModalOpen} 
              expenseDescription={this.props.expense.description}
              handleRemoveExpense={this.onRemove}
              handleCloseModal={this.closeModal}
            />
          </div>
        ) :
        (
          <LoadingPage/>
        )}
      </div>
    );
  };
};


const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: (id) => dispatch(startRemoveExpense({ id }))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);