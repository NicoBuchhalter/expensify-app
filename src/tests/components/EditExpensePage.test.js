import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let startEditExpense, startRemoveExpense, history, wrapper;
beforeEach(() => {
  startEditExpense = jest.fn().mockImplementation(() => Promise.resolve('Edited!'));
  startRemoveExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditExpensePage
      expense={expenses[1]} 
      startEditExpense={startEditExpense}
      startRemoveExpense={startRemoveExpense} 
      history={history} 
    />
  );
});

test('Should render EditExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('Should handle onSubmit', (done) => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
  expect(startEditExpense).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]);
  startEditExpense().then(() => {
    expect(history.push).toHaveBeenLastCalledWith('/');
    done();
  })
});