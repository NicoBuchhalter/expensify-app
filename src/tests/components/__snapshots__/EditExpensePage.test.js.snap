// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render EditExpensePage correctly 1`] = `
<div>
  <div>
    <div
      className="page-header"
    >
      <div
        className="content-container"
      >
        <h1
          className="page-header__title"
        >
          Edit Expense
        </h1>
      </div>
    </div>
    <div
      className="content-container"
    >
      <ExpenseForm
        expense={
          Object {
            "amount": 1090000,
            "createdAt": -345600000,
            "description": "Rent",
            "id": "2",
            "note": "",
          }
        }
        onSubmit={[Function]}
        submitText="Save Expense"
      />
      <button
        className="button--secondary"
        onClick={[Function]}
      >
        Remove Expense
      </button>
    </div>
    <RemoveExpenseModal
      expenseDescription="Rent"
      handleCloseModal={[Function]}
      handleRemoveExpense={[Function]}
      isModalOpen={false}
    />
  </div>
</div>
`;
